lumberdev/sum-it-up
# feat: add api for web scraper, react-query

## WHAT IT DOES:

_Description of the pull request, what changed_
add api for web scraper, react-query

- API endpoint can be found by navigating to route `localhost:3000/api/readability`
- API endpoint accepts `GET` with query param appended `localhost:3000/api/readability?url_resource="valid_url_here"`
- You can also optionally append `&chunk_length={number}` to control the length of chunks returned.
- Output body will be in one of these forms:
```ts
export type DocumentResponseData = {
  url: string;
  title: string;
  byline: string;
  dir: string;
  content: string;
  textContent: string;
  length: number;
  excerpt: string;
  siteName: string;
  chunkedTextContent: Array<string>;
};

export type ErrorMessage = {
  message: string;
};
```

Some important notes:
GPT-3 has a consumption limit of about 4000 tokens. In order to comply with the limitation, the response body of this api includes a property called `chunkedTextContent`, it is currently set process text content into strings with 2000 words each. This should be good enough to feed directly to the GPT model as bulk requests without causing large request issues. If however, this chunk is still too large for the prompt request, you can change the chunk length by requesting a smaller or larger chunk length with the query parameters.

Also, occasionally the API will fail while obtaining the text content from the HTML, it may take somewhere between 1 - 3 tries to get a valid response.

## HOW TO TEST:

- Run the dev server or check the preview link
- Navigate to `/api/readability`
- Append the Query parameter `url_resource="{YOUR ARTICLE URL HERE}"`
- Optionally append the second query parameter `chunk_length={CHUNK LENGTH IN INT}`
- Check out the response object.
